pkg/apis/tensorflow/v1alpha1/types.go

```diff
@@ -0,0 +1,195 @@
+	// TfPort is the port to use for TF services.
+	TfPort *int32 `json:"tfPort,omitempty" protobuf:"varint,1,opt,name=tfPort"`
+	TfReplicaType `json:"tfReplicaType"`
+	// IsDefaultPS denotes if the parameter server should use the default grpc_tensorflow_server
+	IsDefaultPS bool
+}
+
+type TensorBoardSpec struct {
+	//Location of TensorFlow event files
+	LogDir       string           `json:"logDir"`
+	Volumes      []v1.Volume      `json:"volumes"`
+	VolumeMounts []v1.VolumeMount `json:"volumeMounts"`
+	ServiceType  v1.ServiceType   `json:"serviceType"`
+}
+
+type TfJobPhase string
+
+const (
+	TfJobPhaseNone     TfJobPhase = ""
+	TfJobPhaseCreating TfJobPhase = "Creating"
+	TfJobPhaseRunning  TfJobPhase = "Running"
+	TfJobPhaseCleanUp  TfJobPhase = "CleanUp"
+	TfJobPhaseFailed   TfJobPhase = "Failed"
+	TfJobPhaseDone     TfJobPhase = "Done"
+)
+
+type State string
+
+const (
+	StateUnknown   State = "Unknown"
+	StateRunning   State = "Running"
+	StateSucceeded State = "Succeeded"
+	StateFailed    State = "Failed"
+)
+
+type TfJobStatus struct {
+	// Phase is the TfJob running phase
+	Phase  TfJobPhase `json:"phase"`
+	Reason string     `json:"reason"`
+
+	// State indicates the state of the job.
+	State State `json:"state"`
+
+	// ReplicaStatuses specifies the status of each TF replica.
+	ReplicaStatuses []*TfReplicaStatus `json:"replicaStatuses"`
+}
+
+type ReplicaState string
+
+const (
+	ReplicaStateUnknown   ReplicaState = "Unknown"
+	ReplicaStateStarting  ReplicaState = "Starting"
+	ReplicaStateRunning   ReplicaState = "Running"
+	ReplicaStateFailed    ReplicaState = "Failed"
+	ReplicaStateSucceeded ReplicaState = "Succeeded"
+)
+
+type TfReplicaStatus struct {
+	TfReplicaType `json:"tf_replica_type"`
+
+	// State is the overall state of the replica
+	State ReplicaState `json:"state"`
+
+	// ReplicasStates provides the number of replicas in each status.
+	ReplicasStates map[ReplicaState]int
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +resource:path=tfjobs
+
+// TfJobList is a list of TfJobs clusters.
+type TfJobList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	Metadata metav1.ListMeta `json:"metadata,omitempty"`
+	// Items is a list of TfJobs
+	Items []TfJob `json:"items"`
+}
+
+type ControllerConfig struct {
+	// Accelerators is a map from the name of the accelerator to the config for that accelerator.
+	// This should match the value specified as a container limit.
+	// e.g. alpha.kubernetes.io/nvidia-gpu
+	Accelerators map[string]AcceleratorConfig
+
+	// Path to the file containing the grpc server source
+	GrpcServerFilePath string
+}
+
+// AcceleratorVolume represents a host path that must be mounted into
+// each container that needs to use GPUs.
+type AcceleratorVolume struct {
+	Name      string
+	HostPath  string
+	MountPath string
```

Comment:
Just FYI, there is no need to mount GPU manually if we use nvidia device-plugin and nvidia-docker.

Reply:
Some useful information here: https://github.com/kubernetes/kubernetes/issues/54011

Reply:
Do all Kubernetes distributions support device-plugin? As of a couple months ago I still think we needed to use the file path on ACS and some other platforms.

Reply:
Yes, I think device-plugin will be beta (enabled by default) in 1.10, and NVIDIA GPU discovery feature (`alpha.kubernetes.io/nvidia-gpu`) will be deprecated in the future.

Please read this doc for more details: https://github.com/kubernetes/website/pull/6736.